File: src/com/train/algorithm/tree/implementInC++/N-aryTreePreorderTraversal.cpp
//
// Created by 罗斯 on 13/10/2022.
//
#include <vector>
#include <stack>

using namespace std;

class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    vector<int> preorder(Node* root) {
//        if (!root) return {};
//        vector<int> ans;
//        stack<Node*> s;
//        s.push(root);
//        while (!s.empty()) {
//            const Node* node = s.top(); s.pop();
//            ans.push_back(node->val);
//            for (auto it = node->children.rbegin(); it != node->children.rend(); ++it)
//                s.push(*it);
//        }
//        return ans;

        vector<int> res;
        preorder(root, res);
        return res;
    }
private:
    void preorder(Node* root, vector<int>& res){
        if (!root)
            return;
        res.push_back(root->val);
        for (const auto& child:root->children) {
            preorder(child,res);
        }
    }
};

File: src/com/train/algorithm/tree/implementInPython/N-aryTreePreorderTraversal.py
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

class Solution(object):
    def traverse(self, root, res):
        if root is None: return
        res.append(root.val)
        for child in root.children:
            self.traverse(child, res)
    """
    def preorder(self, root):
        res = []
        self.traverse(root, res)
        return res
    """

    def preorder(self, root):
        if root is None:
            return []

        stack = [root]
        output = []

        while stack:
            temp = stack.pop()

            output.append(temp.val)

            stack.extend(temp.children[::-1])

        return output